### ISSUE TITLE
NullReferenceException and InvalidCastException when using Package

### ISSUE THREAD

**Issue by adirh3** [microsoft/CsWinRT#408]
Seems like that some of the PackageManager functionality does not work correctly when using .NET 5 preview 8 and cswinrt- 
I tried running on all available net5.0-windows10.0.xxx versions (Preview 8), I had no issues with Microsoft.Windows.SDK.Contracts

The following code crashes only when **running** on Windows 1809 (10.0.17763) **only**:
```
var packages = new PackageManager().FindPackages();
Package logo = packages.First().Logo; // The logo getter throws NRE for ANY package
```

**The exception:**
```
Failed System.NullReferenceException: Object reference not set to an instance of an object.
   at WinRT.ExceptionHelpers.ThrowExceptionForHR(Int32 hr)
   at ABI.Windows.ApplicationModel.IPackage2.get_Logo()
```

**Another problem is**  (happens on every version)
```
var entries = await package.GetAppListEntriesAsync();
 // This throws InvalidCastException
var accessStreamReference = entries.First().DisplayInfo.GetLogo(new Size(256, 256));
```

**The exception:**
```
Failed entires System.InvalidCastException: Unable to cast object of type 'Windows.Storage.StorageFile' to type 'ABI.Windows.Storage.Streams.IRandomAccessStreamReference'.
   at Windows.Storage.Streams.RandomAccessStreamReference.FromAbi(IntPtr thisPtr)
   at ABI.Windows.Storage.Streams.RandomAccessStreamReference.FromAbi(IntPtr thisPtr)
   at ABI.Windows.ApplicationModel.IAppDisplayInfo.GetLogo(Size size)
```

**Comment by AdamBraden** [comment]
@angelazhangmsft - Relnote for RC2

**Comment by DrusTheAxe** [comment]
> The following code crashes only when running on Windows 1809 (10.0.17763) only

No problem on older or newer releases?

Can you change it to
```
var packages = new PackageManager().FindPackages(); 
var package = packages.First();
var packageFullName = package.Id.PackageFullName
var isBundle = package.IsBundle;
var isFramework = package.IsFramework;
var isOptional = package.IsOptional;
var isResource = package.IsResourcePackage;
Package logo = packages.First().Logo;
```
and share the results of those properties before the .Logo call?


> // The logo getter throws NRE for ANY package

NRE = ?


> Another problem is

Does that fail in the call to GetAppListEntriesAsync() or when later *using* the result? Please provide a more complete example

**Comment by j0shuams** [comment]
@DrusTheAxe 

I tried this same code on my machine, a new release, and it had no trouble. Have not tried on older OS though. 

I did what you mentioned and the properties were all False -- except for .PackageFullName, that one caused a "PackageId doesn't contain definition for PackageFullName" error...

NRE is Null Reference Exception. I have found that an error is thrown earlier, "Parameter 1 is null or undefined." and I think this is causing the Null Reference Exception. I haven't been able to debug into OS code on the VM I'm using (yet) to tell more of what is going on, though. I do know that these errors happen when calling get_Logo_3 on the ABI.Windows.ApplicationModel.IPackage2.Vftbl

I got their second scenario working up through the call to GetAppListEntriesAsync() -- which, to answer your question, does not fail/throw any error. But there is a compiler error with trying to call First() on that result, the types dont line up. 
I have tried calling GetResults() on entries first but that throws a different error -- WinRT - Method was called at an unexpected time. 

+1 A more complete example is needed.

**Comment by DrusTheAxe** [comment]
> I did what you mentioned and the properties were all False

That means it's a Main package (there is no .IsMain property)

> except for .PackageFullName, that one caused a "PackageId doesn't contain definition for PackageFullName" error...

Oops. That should be `.FullName

>I got their second scenario working up through the call to GetAppListEntriesAsync()...compiler error with trying to call First() on that result, the types dont line up

Maybe due to the Async? Does this work in earlier releases of .NETCore? Does it work in with .NET 4.x?

> get_Logo_3 on the ABI.Windows.ApplicationModel.IPackage2.Vftbl

StorageItems used to get cranky if the underlying filesystem artifact doesn't exist. If you know the package full name you can `powershell -c $(Get-AppxPackage -name *something*).InstallLocation to determine its path and make sure that's as expected

**Comment by adirh3** [comment]
@DrusTheAxe @j0shuams 

Clarifying things - just tested it again with NET5 RC1 and the problems still exists - 

Csproj (tried all available win10 versions)
```
<TargetFramework>net5.0-windows10.0.17763</TargetFramework>
```

**Problem 1:**
InvalidCastException when using GetLogo for any package and on every Windows 10 version I tested.
```
PackageManager p = new PackageManager();
var package = p.FindPackages().First();
var entries = await package.GetAppListEntriesAsync();
var randomAccessStreamReference = entries.First().DisplayInfo.GetLogo(new Size(256,256));
```

**Problem 2:**
NullReferenceException when using Logo, this does not happen on my Windows 10 machine (ver 2004) and when using a VM with Windows 10 ver 1809 it happens - 
```
var packages = new PackageManager().FindPackages(); 
Package logo = packages.First().Logo;
```
Also, this happens only on 1809 because it's the earliest version supported, anything above that works fine.

Let me know if you need more information.
Thanks

**Comment by DrusTheAxe** [comment]
> NullReferenceException when using Logo, this does not happen on my Windows 10 machine (ver 2004) and when using a VM with Windows 10 ver 1809 it happens -

Aha! OK, that I understand. This is the same issue as [Question: Would Package.DisplayName be available in WinUI 3 desktop apps? #81](https://github.com/microsoft/ProjectReunion/issues/81)

TL;DR It's a feature addressed in newer Windows (>=19H1)

Package manifest'd localizable properties (DisplayName, PublisherDisplayName, logos, etc) were available if the package is acquired via Package.Current but not via PackageManager.Find*(). This is doc'd e.g. [Package.DisplayName Property](https://docs.microsoft.com/uwp/api/windows.applicationmodel.package.displayname?view=winrt-19041) notes
```
Remarks

Important

This property is only supported for UWP app. If you call this property on an instance of a package that was created by any mechanism other than Package.Current, this property returns an empty string ("").
```
This was addressed in 10.0.18362.0 (aka May 2019 Update aka 19H1) to work regardless how the Package object was created.

In the case of the logo properties I expect if the Package object wasn't created via Package.Current we' resolve the logo's path to `null` or `""` and then try to create a `StorageFolder` with that fails.

The link to [Project Reunion issue #81](https://github.com/microsoft/ProjectReunion/issues/81) describes the issue in more detail. I also experimented a bit with what a Project Reunion style solution would look like. That thread has more details. Does this interest you?

**Comment by adirh3** [comment]
> TL;DR It's a feature addressed in newer Windows (>=19H1)

Thanks for the explanation, it is interesting.
I am using the `PackageManager` for getting information about other packages, so `Package.Current` is of no use to me.
Once`package.GetLogo` is fixed (the first problem), will it also not work in 1809? 

Thanks

**Comment by DrusTheAxe** [comment]
> Once package.GetLogo is fixed (the first problem), will it also not work in 1809?
The change was made in 1903. It won't appear in older releases unless it was serviced downlevel, and servicing tends to have a high bar i.e. most changes don't get backported to older released versions of Windows. I'd guess this didn't meet the bar.

One of Project Reunion's goals is to help provide that sort of backported/downlevel functionality developers would like but Windows' servicing can't justify.

Package's documentation was supposed to be updated to reflect this limitation lifted as of 1903. I'll ping some folks to help get that addressed.

**Comment by j0shuams** [comment]
@adirh3 As I understand it, we are calling the NRE issue closed. 

I have tried to reproduce the Invalid Cast Exception, but cannot get to the GetLogo call. The call to First() is failing -- GetAppListEntriesAsync is returning an empty list. I have tried Iterating over all packages returned by FindPackages() and FindPackagesForUser("") and seeing if any have a non-empty entries list, but no luck. 
Let me know if you have another way to reproduce this, or a suggestion for getting a non-empty list of AppListEntry.

**Comment by adirh3** [comment]
@j0shuams 
Seems like there is another problem that `FindPackages()` and `FindPackagesForUser("")` does not work on .NET 5 RC1 (used to work on NET Core 3.1), should I open issue for it as well?

As for the `GetLogo` this should work for you:
This code find all the installed packages on my machine and crashes on GetLogo:

```
var packageManager = new PackageManager();
var findPackages = packageManager.FindPackagesForUserWithPackageTypes("", PackageTypes.Main);
foreach (var package in findPackages)
{
    try
    {
        var entries = await package.GetAppListEntriesAsync();
        var firstEntry = entries.FirstOrDefault();
        if (firstEntry != null)
        {
            var logo = firstEntry.DisplayInfo.GetLogo(new Size(256, 256));
            Console.WriteLine(logo);
        }
    }
    catch (Exception e)
    {
        Console.WriteLine($"Failed: {e}");
    }
}
```

**Comment by j0shuams** [comment]
That sounds like a Dotnet issue and not a CsWinRT issue, so I think you would open an issue with them. For me, I am using .NET 5 RC2 and ```FindPackages``` and ```FindPackagesForUser("")``` did not cause any problems. 

And thank you for your quick reply, but unfortunately that code did not cause a crash. For me, all of the `firstEntry` were null.

**Comment by j0shuams** [comment]
I see what you mean, I didn't think anything was wrong with FindPackages* because they are returning some packages. Maybe the issue is with GetAppListEntriesAsync() instead, but I'm not ruling out FindPackages* either now. This is confusing, and I'm not sure myself; I'll figure out who owns this.

**Comment by adirh3** [comment]
>This is confusing, and I'm not sure myself; I'll figure out who owns this.

Okay, thank you. Let me know if I need to open issue in the .NET repo.

**Comment by j0shuams** [comment]
@adirh3 Would you try this with .NET RC2 ? We believe that the updates we made and got moved into RC2 fix this error -- @manodasanW stepped back to before the changes we added in RC2 and this repro'd and when he tried it with RC2 it ran fine. 

And since you and Mano are able to use FindPackages* functions, I don't think this is a cswinrt (or dotnet) issue. Maybe it is an issue with the packages I have installed or some kind of API bug in the OS.

**Comment by adirh3** [comment]
> @adirh3 Would you try this with .NET RC2 ? We believe that the updates we made and got moved into RC2 fix this error -- @manodasanW stepped back to before the changes we added in RC2 and this repro'd and when he tried it with RC2 it ran fine.
> 
> And since you and Mano are able to use FindPackages* functions, I don't think this is a cswinrt (or dotnet) issue. Maybe it is an issue with the packages I have installed or some kind of API bug in the OS.

Sure I will test with RC2 and update

**Comment by adirh3** [comment]
@j0shuams 

Seems like RC2 fixes `GetLogo` for me (at least on 2004)
Thanks!

**Comment by j0shuams** [comment]
Awesome! Thank you
